kml: Whitespace in <color> throws exception

Extra whitespace in a <color> tag will throw a NumberFormatException from Color.parseColor

For example:

         <color>ff0098e6 </color> 

It's not immediately obvious from the XML and KML specs whether this is valid, but in any case, it would make sense to trim the color string before attempting to parse it.


A sample KML with this extra whitespace is https://rmgsc.cr.usgs.gov/outgoing/GeoMAC/ActiveFirePerimeters.kml

Here is a stack trace:

Caused by: java.lang.NumberFormatException: Invalid long: "ff0098e6 "
                                                                                 at java.lang.Long.invalidLong(Long.java:124)
                                                                                 at java.lang.Long.parse(Long.java:361)
                                                                                 at java.lang.Long.parseLong(Long.java:352)
                                                                                 at android.graphics.Color.parseColor(Color.java:211)
                                                                                 at com.google.maps.android.data.kml.KmlStyle.setOutlineColor(KmlStyle.java:334)
                                                                                 at com.google.maps.android.data.kml.KmlStyleParser.createLineStyle(KmlStyleParser.java:195)
                                                                                 at com.google.maps.android.data.kml.KmlStyleParser.createStyle(KmlStyleParser.java:55)
                                                                                 at com.google.maps.android.data.kml.KmlContainerParser.setContainerStyle(KmlContainerParser.java:148)
                                                                                 at com.google.maps.android.data.kml.KmlContainerParser.assignPropertiesToContainer(KmlContainerParser.java:91)
                                                                                 at com.google.maps.android.data.kml.KmlContainerParser.createContainer(KmlContainerParser.java:52)
                                                                                 at com.google.maps.android.data.kml.KmlParser.parseKml(KmlParser.java:74)
                                                                                 at com.google.maps.android.data.kml.KmlLayer.<init>(KmlLayer.java:47)

--- a/library/src/androidTest/java/com/google/maps/android/data/kml/KmlStyleTest.java
+++ b/library/src/androidTest/java/com/google/maps/android/data/kml/KmlStyleTest.java
@@ -1,7 +1,9 @@
 package com.google.maps.android.data.kml;
 
 import com.google.android.gms.maps.MapsInitializer;
+import com.google.maps.android.TestUtil;
 
+import org.junit.Assume;
 import org.junit.Before;
 import org.junit.Test;
 
@@ -113,6 +115,10 @@ public class KmlStyleTest {
 
     @Test
     public void testMarkerColorLeadingSpace() {
+        if (TestUtil.isRunningOnTravis()) {
+            Assume.assumeTrue("Skipping KmlStyleTest.testMarkerColorLeadingSpace() - this is expected behavior on Travis CI (#573)", false);
+            return;
+        }
         KmlStyle kmlStyle = new KmlStyle();
         assertNotNull(kmlStyle);
         assertNotNull(kmlStyle.getMarkerOptions());
@@ -122,6 +128,10 @@ public class KmlStyleTest {
 
     @Test
     public void testMarkerColorTrailingSpace() {
+        if (TestUtil.isRunningOnTravis()) {
+            Assume.assumeTrue("Skipping KmlStyleTest.testMarkerColorTrailingSpace() - this is expected behavior on Travis CI (#573)", false);
+            return;
+        }
         KmlStyle kmlStyle = new KmlStyle();
         assertNotNull(kmlStyle);
         assertNotNull(kmlStyle.getMarkerOptions());

--- a/library/src/androidTest/java/com/google/maps/android/data/kml/KmlStyleTest.java
+++ b/library/src/androidTest/java/com/google/maps/android/data/kml/KmlStyleTest.java
@@ -1,12 +1,26 @@
 package com.google.maps.android.data.kml;
 
-import android.graphics.Color;
+import com.google.android.gms.maps.MapsInitializer;
 
+import org.junit.Before;
 import org.junit.Test;
 
-import static org.junit.Assert.*;
+import android.graphics.Color;
+
+import androidx.test.platform.app.InstrumentationRegistry;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
 
 public class KmlStyleTest {
+
+    @Before
+    public void setUp() {
+        MapsInitializer.initialize(InstrumentationRegistry.getInstrumentation().getTargetContext());
+    }
+
     @Test
     public void testStyleId() {
         KmlStyle kmlStyle = new KmlStyle();
@@ -34,14 +48,31 @@ public class KmlStyleTest {
     }
 
     @Test
+    public void testFillColorLeadingSpace() {
+        KmlStyle kmlStyle = new KmlStyle();
+        assertNotNull(kmlStyle);
+        assertNotNull(kmlStyle.getPolygonOptions());
+        kmlStyle.setFillColor(" 000000");
+        int fillColor = Color.parseColor("#000000");
+        assertEquals(fillColor, kmlStyle.getPolygonOptions().getFillColor());
+    }
+
+    @Test
+    public void testFillColorTrailingSpace() {
+        KmlStyle kmlStyle = new KmlStyle();
+        assertNotNull(kmlStyle);
+        assertNotNull(kmlStyle.getPolygonOptions());
+        kmlStyle.setFillColor("000000 ");
+        int fillColor = Color.parseColor("#000000");
+        assertEquals(fillColor, kmlStyle.getPolygonOptions().getFillColor());
+    }
+
+    @Test
     public void testColorFormatting() {
         KmlStyle kmlStyle = new KmlStyle();
         // AABBGGRR -> AARRGGBB.
         kmlStyle.setFillColor("ff579D00");
         assertEquals(Color.parseColor("#009D57"), kmlStyle.getPolygonOptions().getFillColor());
-        // Alpha w/ missing 0.
-        kmlStyle.setFillColor(" D579D00");
-        assertEquals(Color.parseColor("#0D009D57"), kmlStyle.getPolygonOptions().getFillColor());
     }
 
     @Test
@@ -81,9 +112,46 @@ public class KmlStyleTest {
     }
 
     @Test
-    public void testMarkerColor() {
+    public void testMarkerColorLeadingSpace() {
         KmlStyle kmlStyle = new KmlStyle();
         assertNotNull(kmlStyle);
         assertNotNull(kmlStyle.getMarkerOptions());
+        kmlStyle.setMarkerColor(" FFFFFF");
+        // We can't get the marker color to see if it set correctly
+    }
+
+    @Test
+    public void testMarkerColorTrailingSpace() {
+        KmlStyle kmlStyle = new KmlStyle();
+        assertNotNull(kmlStyle);
+        assertNotNull(kmlStyle.getMarkerOptions());
+        kmlStyle.setMarkerColor("FFFFFF ");
+        // We can't get the marker color to see if it set correctly
+    }
+
+    @Test
+    public void testLineColorWithLeadingSpace() {
+        KmlStyle kmlStyle = new KmlStyle();
+        assertNotNull(kmlStyle);
+        assertNotNull(kmlStyle.getPolygonOptions());
+        assertNotNull(kmlStyle.getPolylineOptions());
+        assertEquals(Color.BLACK, kmlStyle.getPolylineOptions().getColor());
+        assertEquals(Color.BLACK, kmlStyle.getPolygonOptions().getStrokeColor());
+        kmlStyle.setOutlineColor(" FFFFFFFF");
+        assertEquals(Color.WHITE, kmlStyle.getPolylineOptions().getColor());
+        assertEquals(Color.WHITE, kmlStyle.getPolygonOptions().getStrokeColor());
+    }
+
+    @Test
+    public void testLineColorWithTrailingSpace() {
+        KmlStyle kmlStyle = new KmlStyle();
+        assertNotNull(kmlStyle);
+        assertNotNull(kmlStyle.getPolygonOptions());
+        assertNotNull(kmlStyle.getPolylineOptions());
+        assertEquals(Color.BLACK, kmlStyle.getPolylineOptions().getColor());
+        assertEquals(Color.BLACK, kmlStyle.getPolygonOptions().getStrokeColor());
+        kmlStyle.setOutlineColor("FFFFFFFF ");
+        assertEquals(Color.WHITE, kmlStyle.getPolylineOptions().getColor());
+        assertEquals(Color.WHITE, kmlStyle.getPolygonOptions().getStrokeColor());
     }
 }

--- a/library/src/androidTest/java/com/google/maps/android/data/kml/KmlStyleTest.java
+++ b/library/src/androidTest/java/com/google/maps/android/data/kml/KmlStyleTest.java
@@ -123,7 +123,7 @@ public class KmlStyleTest {
         assertNotNull(kmlStyle);
         assertNotNull(kmlStyle.getMarkerOptions());
         kmlStyle.setMarkerColor(" FFFFFF");
-        // We can't get the marker color to see if it set correctly
+        assertEquals(Color.WHITE, kmlStyle.mMarkerColor, 1);
     }
 
     @Test
@@ -136,7 +136,7 @@ public class KmlStyleTest {
         assertNotNull(kmlStyle);
         assertNotNull(kmlStyle.getMarkerOptions());
         kmlStyle.setMarkerColor("FFFFFF ");
-        // We can't get the marker color to see if it set correctly
+        assertEquals(Color.WHITE, kmlStyle.mMarkerColor, 1);
     }
 
     @Test

--- a/library/src/main/java/com/google/maps/android/data/kml/KmlStyle.java
+++ b/library/src/main/java/com/google/maps/android/data/kml/KmlStyle.java
@@ -190,8 +190,6 @@ public class KmlStyle extends Style {
      * @param color Fill color for a KML Polygon as a String
      */
     /* package */ void setFillColor(String color) {
-        // Tolerate KML with leading or trailing whitespace in colors
-        color = color.trim();
         // Add # to allow for mOutline color to be parsed correctly
         int polygonColorNum = (Color.parseColor("#" + convertColor(color)));
         setPolygonFillColor(polygonColorNum);
@@ -204,8 +202,6 @@ public class KmlStyle extends Style {
      * @param color Color for a marker
      */
     /* package */ void setMarkerColor(String color) {
-        // Tolerate KML with leading or trailing whitespace in colors
-        color = color.trim();
         int integerColor = Color.parseColor("#" + convertColor(color));
         mMarkerColor = getHueValue(integerColor);
         mMarkerOptions.icon(BitmapDescriptorFactory.defaultMarker(mMarkerColor));
@@ -225,12 +221,15 @@ public class KmlStyle extends Style {
     }
 
     /**
-     * Converts a color format of the form AABBGGRR to AARRGGBB
+     * Converts a color format of the form AABBGGRR to AARRGGBB. Any leading or trailing spaces
+     * in the provided string will be trimmed prior to conversion.
      *
      * @param color Color of the form AABBGGRR
      * @return Color of the form AARRGGBB
      */
     private static String convertColor(String color) {
+        // Tolerate KML with leading or trailing whitespace in colors
+        color = color.trim();
         String newColor;
         if (color.length() > 6) {
             newColor = color.substring(0, 2) + color.substring(6, 8)
@@ -332,8 +331,6 @@ public class KmlStyle extends Style {
      * @param color Outline color for a Polyline and a Polygon represented as a String
      */
     /* package */ void setOutlineColor(String color) {
-        // Tolerate KML with leading or trailing whitespace in colors
-        color = color.trim();
         // Add # to allow for mOutline color to be parsed correctly
         mPolylineOptions.color(Color.parseColor("#" + convertColor(color)));
         mPolygonOptions.strokeColor(Color.parseColor("#" + convertColor(color)));

--- a/library/src/main/java/com/google/maps/android/data/kml/KmlStyle.java
+++ b/library/src/main/java/com/google/maps/android/data/kml/KmlStyle.java
@@ -1,12 +1,12 @@
 package com.google.maps.android.data.kml;
 
-import android.graphics.Color;
-
 import com.google.android.gms.maps.model.BitmapDescriptorFactory;
 import com.google.android.gms.maps.model.MarkerOptions;
 import com.google.android.gms.maps.model.PolygonOptions;
 import com.google.android.gms.maps.model.PolylineOptions;
 import com.google.maps.android.data.Style;
+
+import android.graphics.Color;
 
 import java.util.HashMap;
 import java.util.HashSet;
@@ -187,6 +187,8 @@ public class KmlStyle extends Style {
      * @param color Fill color for a KML Polygon as a String
      */
     /* package */ void setFillColor(String color) {
+        // Tolerate KML with leading or trailing whitespace in colors
+        color = color.trim();
         // Add # to allow for mOutline color to be parsed correctly
         int polygonColorNum = (Color.parseColor("#" + convertColor(color)));
         setPolygonFillColor(polygonColorNum);
@@ -199,6 +201,8 @@ public class KmlStyle extends Style {
      * @param color Color for a marker
      */
     /* package */ void setMarkerColor(String color) {
+        // Tolerate KML with leading or trailing whitespace in colors
+        color = color.trim();
         int integerColor = Color.parseColor("#" + convertColor(color));
         mMarkerColor = getHueValue(integerColor);
         mMarkerOptions.icon(BitmapDescriptorFactory.defaultMarker(mMarkerColor));
@@ -231,10 +235,6 @@ public class KmlStyle extends Style {
         } else {
             newColor = color.substring(4, 6) + color.substring(2, 4) +
                     color.substring(0, 2);
-        }
-        // Maps exports KML colors with a leading 0 as a space.
-        if (newColor.substring(0, 1).equals(" ")) {
-            newColor = "0" + newColor.substring(1, newColor.length());
         }
         return newColor;
     }
@@ -329,6 +329,8 @@ public class KmlStyle extends Style {
      * @param color Outline color for a Polyline and a Polygon represented as a String
      */
     /* package */ void setOutlineColor(String color) {
+        // Tolerate KML with leading or trailing whitespace in colors
+        color = color.trim();
         // Add # to allow for mOutline color to be parsed correctly
         mPolylineOptions.color(Color.parseColor("#" + convertColor(color)));
         mPolygonOptions.strokeColor(Color.parseColor("#" + convertColor(color)));

--- a/library/src/main/java/com/google/maps/android/data/kml/KmlStyle.java
+++ b/library/src/main/java/com/google/maps/android/data/kml/KmlStyle.java
@@ -12,6 +12,8 @@ import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Random;
 
+import androidx.annotation.VisibleForTesting;
+
 /**
  * Represents the defined styles in the KML document
  */
@@ -43,7 +45,8 @@ public class KmlStyle extends Style {
 
     private boolean mPolyRandomColorMode;
 
-    private float mMarkerColor;
+    @VisibleForTesting
+    float mMarkerColor;
 
     /**
      * Creates a new KmlStyle object

